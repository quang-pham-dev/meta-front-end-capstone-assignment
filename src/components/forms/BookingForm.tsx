import React, { useState, FormEvent } from "react";

import { Button } from "@/components/ui/Button";
import { Dropdown } from "@/components/ui/Dropdown";
import { ChooseDateInput } from "@/components/ChooseDateInput";
import { NumberOfGuestsInput } from "@/components/NumberOfGuestInput";

interface BookingFormProps {
  availableTimes: string[];
  onSubmit: (formData: FormData) => void;
}

const BookingForm: React.FC<BookingFormProps> = ({
  availableTimes,
  onSubmit
}) => {
  const [formState, setFormState] = useState({
    date: "",
    time: "",
    guests: "",
    occasion: ""
  });

  const handleInputChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = e.target;
    setFormState((prev) => ({ ...prev, [name]: value }));
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const formData = new FormData(e.currentTarget);
    onSubmit(formData);
  };

  return (
    <div className="booking-form-container">
      <form onSubmit={handleSubmit}>
        <fieldset>
          <ChooseDateInput
            id="book-date"
            name="date"
            value={formState.date}
            onChange={handleInputChange}
            required
            aria-label="Choose date"
          />

          <Dropdown
            id="book-time"
            name="time"
            value={formState.time}
            onChange={handleInputChange}
            required
            options={availableTimes.map((time) => ({
              value: time,
              label: time
            }))}
            label="Choose Time:"
            aria-label="Choose time"
          />

          <NumberOfGuestsInput
            id="book-guests"
            name="guests"
            value={formState.guests}
            onChange={handleInputChange}
            required
            aria-label="Number of guests"
          />

          <Dropdown
            id="book-occasion"
            name="occasion"
            value={formState.occasion}
            onChange={handleInputChange}
            required
            options={[
              { value: "Birthday", label: "Birthday" },
              { value: "Anniversary", label: "Anniversary" }
            ]}
            label="Occasion:"
            aria-label="Choose occasion"
          />

          <Button
            type="submit"
            className="button-receive"
            aria-label="Make Your Reservation"
          >
            Make Your Reservation
          </Button>
        </fieldset>
      </form>
    </div>
  );
};

export default BookingForm;
