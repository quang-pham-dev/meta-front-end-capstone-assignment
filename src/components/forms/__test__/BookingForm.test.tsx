import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import BookingForm from "../BookingForm";
import "@testing-library/jest-dom";

describe("BookingForm", () => {
  it("matches snapshot", () => {
    const mockAvailableTimes = ["17:00", "18:00", "19:00"];
    const mockOnSubmit = vi.fn();

    const { asFragment } = render(
      <BookingForm
        availableTimes={mockAvailableTimes}
        onSubmit={mockOnSubmit}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const mockAvailableTimes = ["17:00", "18:00", "19:00"];
    const mockOnSubmit = vi.fn();

    render(
      <BookingForm
        availableTimes={mockAvailableTimes}
        onSubmit={mockOnSubmit}
      />
    );

    expect(screen.getByLabelText(/choose date/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/choose time/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/number of guests/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/occasion/i)).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /make your reservation/i })
    ).toBeInTheDocument();
  });

  it("submits the form with correct data", () => {
    const mockAvailableTimes = ["17:00", "18:00", "19:00"];
    const mockOnSubmit = vi.fn();

    render(
      <BookingForm
        availableTimes={mockAvailableTimes}
        onSubmit={mockOnSubmit}
      />
    );

    fireEvent.change(screen.getByLabelText(/choose date/i), {
      target: { value: "2023-05-01" }
    });
    fireEvent.change(screen.getByLabelText(/choose time/i), {
      target: { value: "18:00" }
    });
    fireEvent.change(screen.getByLabelText(/number of guests/i), {
      target: { value: "4" }
    });
    fireEvent.change(screen.getByLabelText(/occasion/i), {
      target: { value: "Birthday" }
    });

    fireEvent.click(
      screen.getByRole("button", { name: /make your reservation/i })
    );

    expect(mockOnSubmit).toHaveBeenCalledWith(expect.any(FormData));
  });
});
