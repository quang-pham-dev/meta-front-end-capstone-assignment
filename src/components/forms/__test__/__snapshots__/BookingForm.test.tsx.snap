// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookingForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="booking-form-container"
  >
    <form>
      <fieldset>
        <div
          class="booking-form-field "
        >
          <label
            class="booking-form-label "
            for="book-date"
          >
            Choose Date:
          </label>
          <input
            aria-label="Choose date"
            class="booking-form-input "
            id="book-date"
            name="date"
            required=""
            type="date"
            value=""
          />
        </div>
        <div
          class="booking-form-field "
        >
          <label
            class="booking-form-label "
            for="book-time"
          >
            Choose Time:
          </label>
          <select
            aria-label="Choose time"
            class="booking-form-input "
            id="book-time"
            name="time"
            required=""
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="17:00"
            >
              17:00
            </option>
            <option
              value="18:00"
            >
              18:00
            </option>
            <option
              value="19:00"
            >
              19:00
            </option>
          </select>
        </div>
        <div
          class="booking-form-field "
        >
          <label
            class="booking-form-label "
            for="book-guests"
          >
            Number of guests:
          </label>
          <input
            aria-label="Number of guests"
            class="booking-form-input "
            id="book-guests"
            max="10"
            min="1"
            name="guests"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          class="booking-form-field "
        >
          <label
            class="booking-form-label "
            for="book-occasion"
          >
            Occasion:
          </label>
          <select
            aria-label="Choose occasion"
            class="booking-form-input "
            id="book-occasion"
            name="occasion"
            required=""
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Birthday"
            >
              Birthday
            </option>
            <option
              value="Anniversary"
            >
              Anniversary
            </option>
          </select>
        </div>
        <button
          aria-label="Make Your Reservation"
          class="button button--default button--medium button-receive"
          type="submit"
        >
          Make Your Reservation
        </button>
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
